File: src/algorithms/sets/optimization/set-cover/set-cover.test.ts
import { describe, it, expect } from "vitest";
import { setCover } from "./sources/set-cover.ts?fn";

describe("setCover", () => {
  it("covers the default universe with the expected number of sets", () => {
    const result = setCover(
      [1, 2, 3, 4, 5, 6, 7, 8],
      [
        [1, 2, 3],
        [2, 4],
        [3, 4, 5],
        [5, 6, 7],
        [6, 7, 8],
      ],
    ) as { selectedIndices: number[]; selectedSets: number[][] };
    // Greedy should cover all 8 elements
    const coveredElements = new Set(result.selectedSets.flat());
    expect(coveredElements).toContain(1);
    expect(coveredElements).toContain(8);
    expect(result.selectedSets.length).toBeGreaterThan(0);
    expect(result.selectedSets.length).toBeLessThanOrEqual(5);
  });

  it("returns a single set when one set covers the entire universe", () => {
    const result = setCover([1, 2, 3], [[1, 2, 3], [1], [2]]) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    expect(result.selectedSets.length).toBe(1);
    expect(result.selectedIndices[0]).toBe(0);
  });

  it("covers a universe of disjoint singletons", () => {
    const result = setCover([1, 2, 3], [[1], [2], [3]]) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    const covered = new Set(result.selectedSets.flat());
    expect(covered.has(1)).toBe(true);
    expect(covered.has(2)).toBe(true);
    expect(covered.has(3)).toBe(true);
    expect(result.selectedSets.length).toBe(3);
  });

  it("selects the greediest set first", () => {
    // S0 covers 3 elements, S1 covers 1 — greedy must pick S0 first
    const result = setCover([1, 2, 3, 4], [[1, 2, 3], [4]]) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    expect(result.selectedIndices[0]).toBe(0);
  });

  it("covers all elements when all sets overlap heavily", () => {
    const result = setCover([1, 2, 3, 4, 5], [[1, 2, 3, 4, 5], [1, 2], [3, 4], [5]]) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    const covered = new Set(result.selectedSets.flat());
    for (const element of [1, 2, 3, 4, 5]) {
      expect(covered.has(element)).toBe(true);
    }
  });

  it("returns empty selection when universe is empty", () => {
    const result = setCover(
      [],
      [
        [1, 2],
        [3, 4],
      ],
    ) as { selectedIndices: number[]; selectedSets: number[][] };
    expect(result.selectedIndices.length).toBe(0);
    expect(result.selectedSets.length).toBe(0);
  });

  it("handles a single-element universe covered by the first matching set", () => {
    const result = setCover([7], [[1, 2], [7, 8], [3]]) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    expect(result.selectedIndices.length).toBe(1);
    const covered = new Set(result.selectedSets.flat());
    expect(covered.has(7)).toBe(true);
  });

  it("returns selectedIndices that match the positions of selectedSets", () => {
    const allSets = [
      [1, 2, 3],
      [2, 4],
      [3, 4, 5],
      [5, 6, 7],
      [6, 7, 8],
    ];
    const result = setCover([1, 2, 3, 4, 5, 6, 7, 8], allSets) as {
      selectedIndices: number[];
      selectedSets: number[][];
    };
    for (let pos = 0; pos < result.selectedIndices.length; pos++) {
      const idx = result.selectedIndices[pos]!;
      expect(result.selectedSets[pos]).toEqual(allSets[idx]);
    }
  });
});

File: src/algorithms/sets/optimization/set-cover/SetCover_test.cpp
#define TESTING
#include "sources/SetCover.cpp"
#include <cassert>
#include <unordered_set>
#include <iostream>

int main() {
    // covers default universe
    std::vector<std::vector<int>> sets1 = {{1,2,3},{2,4},{3,4,5},{5,6,7},{6,7,8}};
    auto result1 = setCover({1,2,3,4,5,6,7,8}, sets1);
    std::unordered_set<int> covered1;
    for (int idx : result1.selectedIndices) {
        for (int elem : sets1[idx]) covered1.insert(elem);
    }
    assert(covered1.count(1) && covered1.count(8));
    assert(!result1.selectedIndices.empty() && result1.selectedIndices.size() <= 5);

    // single set covers universe
    std::vector<std::vector<int>> sets2 = {{1,2,3},{1},{2}};
    auto result2 = setCover({1,2,3}, sets2);
    assert(result2.selectedIndices.size() == 1);
    assert(result2.selectedIndices[0] == 0);

    // greediest set selected first
    std::vector<std::vector<int>> sets4 = {{1,2,3},{4}};
    auto result4 = setCover({1,2,3,4}, sets4);
    assert(result4.selectedIndices[0] == 0);

    // empty universe returns empty selection
    auto result5 = setCover({}, {{1,2},{3,4}});
    assert(result5.selectedIndices.empty());

    // selected indices match selected sets
    std::vector<std::vector<int>> allSets = {{1,2,3},{2,4},{3,4,5},{5,6,7},{6,7,8}};
    auto result6 = setCover({1,2,3,4,5,6,7,8}, allSets);
    for (size_t pos = 0; pos < result6.selectedIndices.size(); pos++) {
        assert(result6.selectedSets[pos] == allSets[result6.selectedIndices[pos]]);
    }

    std::cout << "All tests passed!" << std::endl;
    return 0;
}
